source/common/network/connection_impl.cc

```diff
@@ -342,7 +352,12 @@ void ConnectionImpl::raiseEvent(ConnectionEvent event) {
   }
 }
 
-bool ConnectionImpl::readEnabled() const { return read_enabled_; }
+bool ConnectionImpl::readEnabled() const {
+  // Calls to readEnabled on a closed socket are considered to be an error.
```

Comment:
Good call.
I wonder if the readEnabled loop in ConnectionManagerImpl::doEndStream could be subject to this in any weird corner case?  I _think_ it should be fine because if the connection gets closed it should be in DrainState::Closing but it might be worth adding an open check in that loop as well

Reply:
I think that the "drain_state_ != DrainState::Closing" check in ConnectionManagerImpl::doEndStream is equivalent.  Or at least I don't see any tests failing due to me adding this ASSERT.